crates/core-subsystem/core-plugin-shared/src/serializable_system.rs

```diff
@@ -14,30 +14,176 @@ use super::{
+/// should fail at a later stage since a matching loader won't be found.
+#[derive(Serialize, Deserialize, Debug)]
+struct Header {
+    builder_version: String,
+    ir_version: String,
+    plugins: Vec<String>,
+}
+
+impl Header {
+    fn new(plugins: Vec<String>) -> Header {
+        let version = env!("CARGO_PKG_VERSION").to_string();
+        Header {
+            builder_version: version.clone(),
+            ir_version: version,
+            plugins,
+        }
+    }
+
+    fn check_header(&self, header: Header) -> Result<(), String> {
+        if self.ir_version != header.ir_version {
+            return Err(format!(
+                "Version for this file {0} does not match current version {1}",
+                header.ir_version, self.ir_version
+            ));
+        }
+        if self.builder_version != header.builder_version {
+            return Err(format!(
+                "Builder version for this file {0} does not match current version {1}",
+                header.builder_version, self.builder_version
+            ));
+        }
+        Ok(())
+    }
+}
+
 impl SystemSerializer for SerializableSystem {
     type Underlying = Self;
 
     fn serialize(&self) -> Result<Vec<u8>, ModelSerializationError> {
-        bincode::serialize(self).map_err(ModelSerializationError::Serialize)
+        serialize_header_and_system(&Header::new(vec![]), self)
     }
 
     fn deserialize_reader(
-        reader: impl std::io::Read,
+        mut reader: impl std::io::Read,
     ) -> Result<Self::Underlying, ModelSerializationError> {
-        bincode::deserialize_from(reader).map_err(ModelSerializationError::Deserialize)
+        // TODO: ModelSerializationError should not be dependent on bincode errors since
+        // it is used by subsystem serializers which may use other formats.
+        fn error(msg: &str, io_error: Option<std::io::Error>) -> ModelSerializationError {
+            let msg = match io_error {
+                Some(e) => format!("{msg}: {e}"),
+                None => msg.to_string(),
+            };
+            ModelSerializationError::Deserialize(bincode::ErrorKind::Custom(msg).into())
+        }
+        {
+            // Check the file prefix
+            let mut prefix: [u8; PREFIX_TAG_LEN] = [0; PREFIX_TAG_LEN];
+            reader
+                .read_exact(&mut prefix)
+                .map_err(|e| error("Failed to read exograph prefix", Some(e)))?;
+
+            if prefix != PREFIX_TAG {
+                return Err(error("Invalid exograph file prefix", None));
+            }
+        }
+        let header_len = {
+            let mut header_len: [u8; 8] = [0; 8];
+            reader
+                .read_exact(&mut header_len)
+                .map_err(|e| error("Failed to read exograph header size", Some(e)))?;
+            usize::from_le_bytes(header_len)
+        };
+        // Check the header
+        let mut remaining = Vec::with_capacity(10000);
```

Comment:
This will fail if exo_ir file is larger that 10000 bytes?

I believe if you pass `reader` (which will now point at the first byte of the "system" portion) to `bincode::deserialize_from`, it will work without any length limit.

Reply:
No, `with_capacity` just ensures a minimum capacity, rather than repeatedly re-allocating from an empty array. It's probably unnecessary, but I scanned through the `ir` files in the integration tests and this seened like a good starting point.

Reply:
At this point it is pointing to the header and the remaining system, not just the system.

Reply:
Ah, I see. Oops!

Reply:
I guess, to point to the start of the system, we would need to read the `header_len` bytes. Then passing the reader onwards will work. 

Probably not a big deal, but I wonder if bincode reads and deserializes in chunks (thus reducing peak memory usage).

Reply:
Hmm. You might be right with the failure - I guess the vector can't reallocate when the reader is reading into it.

Reply:
Changed to read an intermediate `vec![0_u8; header_len]` for the header before passing to bincode as suggested.